### PR TITLE
Transpose binary

### PR DESCRIPTION
**Pull request by defei-coder**
Thanks for your contribution and we appreciate it a lot. 

## 1. Motivation

Add new operator transpose.

## 2. Modification

Please briefly describe what modification is made in this pull request, and indicate where to make the modification.

## 3. Test Report

If you want to know how to do operator testing, you can see [GTest-User-Guide-zh](../docs/GTest-User-Guide-zh.md).

#### 3.1.1 Accuracy Acceptance Standard

For static threshold standard details, see: [MLU-OPS Accuracy Acceptance Standard](../docs/MLU-OPS-Accuracy-Acceptance-Standard.md).

- [ ] diff1, diff1 <= 3e-3
- [ ] diff2, diff2 <= 3e-3

#### 3.1.2 Operator Scheme checklist

|     No.      |           Details            |      Check Results       |
|----------------|---------------------------|---------------------|
|        1       |          Supported hardware         | MLU270 <br> MLU290 <br>MLU370|
|        2       |          Job types          |    block <br> U1 <br> U4    |
|        3       |         Layouts            |  NHWC 、NCHW、ARRAY etc    |
|        4       |         Whether multi-dimensions are supported              |                |
|        5       |          Whether element zero is supported             |               |
|        6       |         Data type(half/float)       |         half / float etc           |
|        7      |        Whether there is size limit           |             |

#### 3.1.3 New Feature Test

If you have checked the following items, please tick the relevant box.

- [ ] Data type test
- [ ] Multidimensional tensor test
- [ ] Layout test
- [ ] Different size/integer remainder end segment/alignment misalignment test
- [ ] Zero dimensional tensor test/zero element test
- [ ] stability test
- [ ] Multiple platform test
- [ ] Gen_case module test
- [ ] Nan/INF tests 
- [ ] Bug fix tests
- [ ] For memory leak check details, see[GTest-User-Guide-zh](../docs/GTest-User-Guide-zh.md).
- [ ] For code coverage check details, see: [GTest-User-Guide-zh](../docs/GTest-User-Guide-zh.md).
- [ ] For I/O calculation efficiency check details, see: [MLU-OPS Performance Acceptance Standard](../docs/MLU-OPS-Performance-Acceptance-Standard.md).

#### 3.1.4 Parameter Check

When a new operator is submitted, the test points are given and the test results are stated.

| Test Point         | Acceptance Standard | Test Result (Error Message) |
| -------------- | -------- | -------------------- |
| Whether it conforms to the operator restriction | Normal error |                      |
| Whether illegal parameters are passed  | Normal error |                      |

### 3.2 Performance Test

See [MLU-OPS Performance Acceptance Standard](../docs/MLU-OPS-Performance-Acceptance-Standard.md) for details.

Platform ：MLU270

|Operator|Mlu_hardware_time(us)|Mlu_interface_time(us)|Mlu_io_efficiency|Mlu_compute_efficiency|Mlu_workwpace_size(Bytes)|Data_type|Shape|
|-----|----|----|----|----|----|------|-----|
|op_name|   |    |     |    |    |    |     |
|op_name|   |    |     |    |    |    |     |

Platform ：MLU290

|Operator|Mlu_hardware_time(us)|Mlu_interface_time(us)|Mlu_io_efficiency|Mlu_compute_efficiency|Mlu_workwpace_size(Bytes)|Data_type|Shape|
|-----|----|----|----|----|----|------|-----|
|op_name|   |    |     |    |    |    |     |
|op_name|   |    |     |    |    |    |     |

Platform：MLU370

|Operator|Mlu_hardware_time(us)|Mlu_interface_time(us)|Mlu_io_efficiency|Mlu_compute_efficiency|Mlu_workwpace_size(Bytes)|Data_type|Shape|
|-----|----|----|----|----|----|------|-----|
|op_name|   |    |     |    |    |    |     |
|op_name|   |    |     |    |    |    |     |

### 3.3 Summary Analysis

Please give a brief overview here, if you want to note and summarize the content.

### CODE REVIEW

File: bangc-ops/mlu_op.h [hunk 1]
```diff
@@ -2871,6 +2871,327 @@ mluOpStatus_t MLUOP_WIN_API mluOpExpand(
...
+ *   queues in the transpose operation. For detailed information,
+ *   see ::mluOpHandle_t.
+ * @param[in]  x_desc
+ *   Input. The descriptor of the input tensor. For detailed information,
+ *   see ::mluOpTensorDescriptor_t.
+ * @param[out] desc
+ *   Input. The descriptor of the transpose operation. For detailed information,
+ *   see ::mluOpTransposeDescriptor_t.
+ * @param[out] size
+ *   Output. A host pointer to the returned size of the extra workspace in bytes that is used in
```
Comment (wushaoqiang2168) [inline bangc-ops/mlu_op.h:2992]: 去掉 Input 、Output 字样

**Review by wushaoqiang2168 — APPROVED** [review]

File: bangc-ops/mlu_op.h [hunk 2]
```diff
@@ -3039,6 +3039,247 @@ mluOpStatus_t MLUOP_WIN_API mluOpFill(
...
+ *  and call the ::mluOpSetTransposeDescriptor function to set the information of
+ *  transpose operation to the descriptor. Also, you need to destroy the Cambricon MLUOP context
+ *  at the end with the ::mluOpDestroyTransposeDescriptor function.
+ */
+typedef struct mluOpTransposeStruct *mluOpTransposeDescriptor_t;
+
+// Group:Transpose
+/*!
+ * @brief Creates a descriptor pointed by \b desc for a transpose operation,
+ *        and allocated memory for holding the information about the transpose operation.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:3056]: allocated > allocates

File: bangc-ops/mlu_op.h [hunk 3]
```diff
@@ -3039,6 +3039,247 @@ mluOpStatus_t MLUOP_WIN_API mluOpFill(
...
+ *
+ * The information is defined in ::mluOpTransposeDescriptor_t. For more information
+ * about descriptor, see "Cambricon MLUOP user Guide".
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:3059]: Cambricon BANGC OPS User Guide 用户手册名称是这个

File: bangc-ops/mlu_op.h [hunk 4]
```diff
@@ -3039,6 +3039,247 @@ mluOpStatus_t MLUOP_WIN_API mluOpFill(
...
+ *
+ * @par Example
+ * - None.
+ */
+mluOpStatus_t MLUOP_WIN_API mluOpCreateTransposeDescriptor(mluOpTransposeDescriptor_t *desc);
+
+// Group:Transpose
+/*!
+ * @brief Initializes the transpose descriptor \b desc that is previously created
+ * with the ::mluOpCreateTransposeDescriptor function, and set the information
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:3086]: set > sets

File: bangc-ops/mlu_op.h [hunk 5]
```diff
@@ -3039,6 +3039,247 @@ mluOpStatus_t MLUOP_WIN_API mluOpFill(
...
+mluOpStatus_t MLUOP_WIN_API mluOpDestroyTransposeDescriptor(mluOpTransposeDescriptor_t desc);
+
+// Group:Transpose
+/*!
+ * @brief Returns in \b size the size of the MLU memory that is used as an extra workspace
+ * to optimize the transpose operation.
+ *
+ * The size of extra workspace is based on the given information of the transpose operation,
+ * including the input tensor descriptor \b x_desc and transpose descriptor \b desc.
+ * For more information about the workspace, see "Cambricon MLUOP User Guide".
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:3147]: Cambricon BANGC OPS User Guide 用户手册名称是这个

File: bangc-ops/mlu_op.h [hunk 6]
```diff
@@ -3039,6 +3039,247 @@ mluOpStatus_t MLUOP_WIN_API mluOpFill(
...
+ *   Pointer to the MLU memory that stores the input tensor.
+ * @param[out] y_desc
+ *   The descriptor of the output tensor. For detailed information,
+ *   see ::mluOpTensorDescriptor_t.
+ * @param[out] y
+ *   Pointer to the MLU memory that stores the output tensor.
+ * @param[in] workspace
+ *   Pointer to the MLU memory that is used as an extra workspace for the
+ *   transpose operation. For more information about workspace,
+ *   see "Cambricon MLUOP User Guide".
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:3214]: 手册名称改一下，见上面

File: bangc-ops/mlu_op.h [hunk 7]
```diff
@@ -3039,6 +3039,247 @@ mluOpStatus_t MLUOP_WIN_API mluOpFill(
...
+ *
+ * @par Scale Limitation
+ * - The \b x, \b y and \b permute have the same shape.
+ * - The dimension size of \b x, \b y and \b permute should be less than or equal to MLUOP_DIM_MAX.
+ * - The \b permute i-th dimension is in the range [0,...n-1], where n is the rank of the \b x.
+ * - The \b y i-th dimension will correspond to \b x permute[i]-th dimension.
+ * - The process of computing, the copy times of memcpy should be less than 65536.
+ *
+ * @par Formula
+ * - See "Transpose Operator" section in "Cambricon MLUOP User Guide" for details.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:3231]: 手册名称改一下

File: bangc-ops/mlu_op.h [hunk 8]
```diff
@@ -3039,6 +3039,247 @@ mluOpStatus_t MLUOP_WIN_API mluOpFill(
...
+
+ * @par Return
+ * - ::MLUOP_STATUS_SUCCESS, ::MLUOP_STATUS_BAD_PARAM
+ *
+ * @par Scale Limitation
+ * - The \b x, \b y and \b permute have the same shape.
+ * - The dimension size of \b x, \b y and \b permute should be less than or equal to MLUOP_DIM_MAX.
+ * - The \b permute i-th dimension is in the range [0,...n-1], where n is the rank of the \b x.
+ * - The \b y i-th dimension will correspond to \b x permute[i]-th dimension.
+ * - The process of computing, the copy times of memcpy should be less than 65536.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:3228]: 计算过程和复制次数必须少于65536，计算过程少于65536是什么意思呢？

File: bangc-ops/mlu_op.h [hunk 9]
```diff
@@ -3039,6 +3039,247 @@ mluOpStatus_t MLUOP_WIN_API mluOpFill(
...
+ *
+ * @par Formula
+ * - See "Transpose Operator" section in "Cambricon MLUOP User Guide" for details.
+ *
+ * @par Data Type
+ * - This function supports the following data types for input tensor \b x and
+ *   output tensor \b y.
+ *   <b>Note that the data type of input tensor and output tensor should be same.</b>
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:3236]: 这句是否可以移到下面的@note字段

另外这句话改为The data types of input tensor and output tensor should be the same.

File: bangc-ops/mlu_op.h [hunk 10]
```diff
@@ -3039,6 +3039,247 @@ mluOpStatus_t MLUOP_WIN_API mluOpFill(
...
+ *   - input tensor: uint8, int8, uint16, int16, uint32, int31, int32, uint64, int64, bool, half,
+ *     float, complex_half, complex_float.
+ *   - output tensor: uint8, int8, uint16, int16, uint32, int31, int32, uint64, int64, bool, half,
+ *     float, complex_half, complex_float.
+ *
+ * @par Data Layout
+ * - The dimension of input tensor should be less than or equal to 8-dimension.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:3243]: 8-dimension > 8

**Review by GraceGuanH — COMMENTED** [review]
Please see the comments

File: bangc-ops/mlu_op.h (hunk 9 above)
Comment (defei-coder) [reply bangc-ops/mlu_op.h:3236]: 这里是想体现数据类型的限制，我修改下表达吧

**Review by GraceGuanH — APPROVED** [review]
Comments are all changed.

File: bangc-ops/kernels/transpose/transpose_host.mlu [hunk 11]
```diff
@@ -0,0 +1,3366 @@
+/*************************************************************************
+ * Copyright (C) [2022] by Cambricon, Inc.
```
Comment (WangYuan-cn) [inline bangc-ops/kernels/transpose/transpose_host.mlu:2]: bangc-ops/kernels/transpose/transpose_host.cpp

**Review by WangYuan-cn — APPROVED** [review]

**Pull request closed without merging**